src/validators.test.ts

```diff
@@ -66,3 +68,131 @@ describe('length', () => {
     expect(got).toEqual(expectedError);
   });
 });
+
+ // Pattern validator
+describe('pattern', () => {
+  it('returns error', () => {
+    const val = 'z';
+    const option = /^[a-b]/;
+    const expectedError = `The value ${val} doen't match the pattern ${option}`;
+    expect(pattern(option)(val)).toEqual(expectedError);
+  });
+  it('returns undefined', () => {
+    expect(pattern(/^[a-z0-9]*$/)('12ba')).toEqual(undefined);
+  });
+  it('considers errorMsg', () => {
+    const errorMsg = 'Custom error.';
+    expect(pattern(/^[a-z0-9]*$/, errorMsg)('abc_123')).toEqual(errorMsg);
+  });
+})
+
+ // minlength validator
+describe('minlength', () => {
+  it('returns error', () => {
+    const expectedError = `Should be at least ${5} characters long`;
+    expect(minlength(5)("suku")).toEqual(expectedError);
+  });
+  it('returns undefined', () => {
+    expect(minlength(5)("summit")).toEqual(undefined);
+  });
+  it('considers errorMsg', () => {
+    const errorMsg = 'Custom error.';
+    expect(minlength(6, errorMsg)("suku")).toEqual(errorMsg);
+  });
+})
+
+ // maxlength validator
+describe('maxlength', () => {
+  it('returns error', () => {
+    const len = 3;
+    const expectedError = `Should be no longer than ${len} characters`;
+    expect(maxlength(len)("suku")).toEqual(expectedError);
+  });
+  it('returns undefined', () => {
+    expect(maxlength(6)("summit")).toEqual(undefined);
+  });
+  it('considers errorMsg', () => {
+    const errorMsg = 'Custom error.';
+    expect(maxlength(3, errorMsg)("suku")).toEqual(errorMsg);
+  });
+})
+
+ // min validator
+describe('min', () => {
+  it('returns error', () => {
+    const val = 12
+    const minVal = 15;
+    const expectedError = `The entered value: ${val} should be at greater than ${minVal}`;
+    expect(min(minVal)(val)).toEqual(expectedError);
+  });
+  it('returns undefined', () => {
+    expect(min(15)(17)).toEqual(undefined);
+  });
+  it('considers errorMsg', () => {
+    const errorMsg = 'Custom error.';
+    expect(min(6, errorMsg)(5)).toEqual(errorMsg);
+  });
+})
+
+ // max validator
+describe('max', () => {
+  it('returns error', () => {
+    const val = 15;
+    const maxVal = 12;
+    const expectedError = `The entered value: ${val} should be smaller than ${maxVal}`;
+    expect(max(maxVal)(val)).toEqual(expectedError);
+  });
+  it('returns undefined', () => {
+    expect(max(6)(4)).toEqual(undefined);
+  });
+  it('considers errorMsg', () => {
+    const maxVal = 3;
+    const errorMsg = `Number smaller than ${maxVal} is expected`;
+    expect(max(maxVal, errorMsg)(5)).toEqual(errorMsg);
+  });
+})
+
+ // date validator
+describe('validateDate', () => {
```

Comment:
We might not need this. The date `decoder` will make this validation. I just you to move this logic to the date validator. At the moment, its implementation is wrong. It should take an ISO string and
convert it to date and validate at the same time.